### ISSUE TITLE
explanation for lstm imdb

### ISSUE THREAD

**Issue by vinayakumarr** [tflearn/tflearn#98]
could you please explain the below code what you used in imdb lstm classification example

# Network building
net = tflearn.input_data([None, 200])
net = tflearn.embedding(net, input_dim=20000, output_dim=128)
net = tflearn.lstm(net, 128)
net = tflearn.dropout(net, 0.5)
net = tflearn.fully_connected(net, 2, activation='softmax')
net = tflearn.regression(net, optimizer='adam',
                         loss='categorical_crossentropy')

# Training
model = tflearn.DNN(net, clip_gradients=0., tensorboard_verbose=0)
model.fit(trainX, trainY, validation_set=(testX, testY), show_metric=True,
          batch_size=128)

**Comment by aymericdamien** [comment]
In this example:
IMDB dataset are user reviews classified into positive or negative review. 
First, it convert words to ids (assign a unique id to every word from the dataset), so a sentence is represented as a list of ids (integers).
Then, we embed that list for representing every id as a vector (that will learn relations between ids, this embedding will be learned during training too) so instead of a list of ids, we get a list of vectors.
Then we apply LSTM network to classify these sequences.

**Comment by vinayakumarr** [comment]
with my data set, (sample is given below)

3 3 373 27 9 615 9 16 10 34 0 8 0 199 65917 1319 122 402 319 183
3 3 77 12 4 66 4 3 0 5 0 14 3 50 106 139 38 164 53 109
3 3 86 6 2 6 2 0 0 1 0 25 0 4 284 77888 19 66 11 25

The first column is the target which has 9 classes and around 1803 features

with my example each cell in a row is assigned to an id, so a feature set is represented as a list of ids(integers).
Then, we embed that list for representing every id as a vector (that will learn relations between ids, this embedding will be learned during training too) so instead of a list of ids, we get a list of vectors.
Then we apply LSTM network to classify these sequences. 

Is that right?

**Comment by aymericdamien** [comment]
Yes, it seems right, you can have a try!

**Comment by vinayakumarr** [comment]
I already normalized my data and able to run the code successfully but the accuracy is showing only 40%. Also I am currently using only training data (splitting training data into train, test and validation). But i would like to classify the test data. How to do this? and also I would like to implement lstm for my data set to test with different network typologies as one memory block with one cell, two memory block with two cell,three memory block with three cell and four memory block with four cell. How to implement this?

**Comment by vinayakumarr** [comment]
I have data set which has 2000 features and before applying lstm i would like to apply random projection and PCA. Is it possible by using tflearn. If so could you please provide a basic example for random projection and PCA (for 2000 features)

**Comment by vinayakumarr** [comment]
I also would like to know which classifier you used in imdb lstm example (for ex softmax).

**Comment by aymericdamien** [comment]
To reduce your dimension, you can start by simply only consider the X events that occur most, and set a similar id (like 0) to all other rarer events, so in this case your embedding dimension will be X. Then according to performances, you may want to try different values for X.
You do not explicitly need validation set, so you can just cut your dataset into train and test.

**Comment by vinayakumarr** [comment]
Which architecture you are using it for imdb example (for ex two memory block with one cell)?
Also I would like to implement lstm for my data set to test with different network typologies as one memory block with one cell, two memory block with two cell,three memory block with three cell and four memory block with four cell. How to implement this?

**Comment by aymericdamien** [comment]
IMDB example is using a single LSTM layer, you can check here: 
https://github.com/tflearn/tflearn/blob/master/tflearn/layers/recurrent.py#L96. It is a simple implementation of LSTM, the cell is defined here: https://github.com/tflearn/tflearn/blob/master/tflearn/layers/recurrent.py#L510.

**Comment by vinayakumarr** [comment]
Thanks. how to write a code for imdb example for the follwoing network toplogies
1. one memory block with one cell
2. two memory block with two cell
3. three memory block with three cell
4. four memory block with four cell
Is it possible to create the toplogies using tflearn?. Please give an example.

**Comment by aymericdamien** [comment]
You can write as follow:
```
# 4 LSTM Layers
net = tflearn.lstm(net, 64, return_seq=True)
net = tflearn.lstm(net, 64, return_seq=True)
net = tflearn.lstm(net, 64, return_seq=True)
net = tflearn.lstm(net, 64, return_seq=False)
```

**Comment by vinayakumarr** [comment]
if i use 
net = tflearn.lstm(net, 64, return_seq=True)

whic meeans one memory block with one cell

and if i use 

net = tflearn.lstm(net, 64, return_seq=True)
net = tflearn.lstm(net, 64, return_seq=False)

which means two memory blocks with two cell.  Is that right

**Comment by aymericdamien** [comment]
Yes

**Comment by vinayakumarr** [comment]
Thanks,I am able to run imdb lstm example for my data set but accuracy is too low and i need to improve this. Once i am able to get good accuracy, I will try contributing to the project as well as soon as I finish the stuff that I am working on. I would like to know the below parametrs or each line explanation with corresponding to my dataset.I already mentioned my data set is 9 classes and around 1803 features. 
In imdb you used 
maxlen=200  //standard length  What will be in my case?
nb_classes=2   // number of classes and in my case 10
net = tflearn.input_data([None, 200])  // 200 is maxlen 
net = tflearn.embedding(net, input_dim=20000, output_dim=128) // i understood from #8 and also i understood embedding is not required for my data set.
net = tflearn.embedding(net, input_dim=20000, output_dim=128)  // input_dim=20000 dictionary size and in my case np.max(trainX)+1 but np.max(trainX)+1 is returning 2 as output is that correct and what will be output_dim for my dataset
net = tflearn.lstm(net, 128) // what is 128 and what will be for my dataset
net = tflearn.dropout(net, 0.5)   // what is 0.5
net = tflearn.fully_connected(net, 2, activation='softmax')  // 2 number of classes and is there any other activation metods are availble. if who can you please specify
net = tflearn.regression(net, optimizer='adam', loss='categorical_crossentropy') //what this line exactly does
model = tflearn.DNN(net, clip_gradients=0., tensorboard_verbose=0) // what is clip_gradients=0 and tensorboard_verbose=0 because i have seen tensorboard_verbose=3 in #8 
model.fit(trainX, trainY, validation_set=(testX, testY), show_metric=True,batch_size=128) // what is show_metric=True and batch_size=128 and what will be for my data set. i have seen 2 other extra parameters n_epoch=2,snapshot_epoch=True in #8. what is this one? shall i include it in my code also.
